backend/asmlink.ml

```diff
@@ -236,12 +236,28 @@ let make_globals_map units_list =
     interfaces
     defined
 
-let make_startup_file ~ppf_dump units_list =
-  let compile_phrase p = Asmgen.compile_phrase ~ppf_dump p in
+let make_startup_file ~ppf_dump ~filename units_list =
   Location.input_name := "caml_startup"; (* set name of "current" input *)
-  Compilenv.reset "_startup";
-  (* set the name of the "current" compunit *)
-  Emit.begin_assembly ();
+  Compilenv.reset "_startup"; (* set the name of the "current" compunit *)
+  let asm_directives =
+    if !Clflags.debug then
+      Some (Asmgen.build_asm_directives ())
+    else
+      None
+  in
+  Emit.begin_assembly (fun () ->
+    Option.iter
+      (fun (module Asm_directives : Asm_targets.Asm_directives_intf.S) ->
+        Asm_targets.Asm_label.initialize ~new_label:Cmm.new_label;
+        Asm_directives.initialize ())
+      asm_directives
+  );
+  let dwarf =
+    Option.map
+      (fun asm_directives -> Asmgen.build_dwarf ~asm_directives filename)
+      asm_directives
```

Comment:
The above sections of code are very similar to the normal case (further up) and should probably be factored out.  Also there needs to be a guard to set `asm_directives` to `None` when the target is not `amd64`.

Reply:
Added a guard.
Not sure I can refactor it better as [build_dwarf] must be called after [Asm_directives.initialize] I believe.